core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java

```diff
@@ -163,18 +163,56 @@ public void addConfigRoot(final PackageElement pkg, TypeElement clazz) {
             }
 
             final String fileName = computeExtensionDocFileName(member);
-            final List<ConfigDocItem> previousExtensionConfigDocKeys = foundExtensionConfigurationItems.get(fileName);
+            final List<ConfigDocItem> existingConfigDocItems = foundExtensionConfigurationItems.get(fileName);
 
-            if (previousExtensionConfigDocKeys == null) {
+            if (existingConfigDocItems == null) {
                 foundExtensionConfigurationItems.put(fileName, configDocItems);
             } else {
-                previousExtensionConfigDocKeys.addAll(configDocItems);
+                // merge two config doc items list
+                mergeConfigItemsIntoExistingOnes(existingConfigDocItems, configDocItems);
             }
         }
 
         return foundExtensionConfigurationItems;
     }
 
+    private void mergeConfigItemsIntoExistingOnes(List<ConfigDocItem> existingConfigDocKeys,
+            List<ConfigDocItem> configDocItems) {
+        for (ConfigDocItem configDocItem : configDocItems) {
+            if (configDocItem.isConfigKey()) {
+                existingConfigDocKeys.add(configDocItem);
+            } else {
+                ConfigDocSection configDocSection = configDocItem.getConfigDocSection();
+                boolean configSectionMerged = mergeSectionIntoPreviousExistingConfigItems(configDocSection,
+                        existingConfigDocKeys);
+                if (!configSectionMerged) {
+                    existingConfigDocKeys.add(configDocItem);
```

Comment:
Why don't you do that in the method itself?

Reply:
The method being recursive, this will help to preserve section ordering. 

Btw, now that I think about it, I should have added test cases for this method.